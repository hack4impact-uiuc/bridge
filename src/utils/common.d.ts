declare module "@hack4impact-uiuc/bridge/common" {
  import { Props, ReactType } from "react";
  import {
    ColorProps,
    SpaceProps,
    LayoutProps,
    TypographyProps as StyledTypographyProps,
    BordersProps as StyledBordersProps,
    BoxShadowProps,
  } from "styled-system";

  export interface BaseProps extends Props<any> {
    as?: ReactType;
    className?: string;
  }
  export interface CommonProps
    extends BaseProps,
      ColorProps,
      SpaceProps,
      LayoutProps {}

  export interface TypographyProps extends BaseProps, StyledTypographyProps {}

  export interface BorderProps extends StyledBordersProps, BoxShadowProps {}
}
