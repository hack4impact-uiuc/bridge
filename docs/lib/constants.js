export const COMPONENTS_URL = '/components';
export const DESIGN_URL = '/design';
