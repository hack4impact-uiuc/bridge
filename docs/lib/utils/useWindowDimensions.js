import { useState, useEffect } from 'react';

function getWindowDimensions() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      const { innerWidth: width, innerHeight: height } = window;
      setWindowDimensions({ width, height });
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}
