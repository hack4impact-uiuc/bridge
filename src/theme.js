/* eslint-disable prefer-destructuring */

function fontStack(fonts) {
  return fonts.map((font) => (font.includes(' ') ? `"${font}"` : font)).join(', ');
}

const DEFAULT_SHADOW = '0 1px 6px 2px rgba(0,0,0,0.15)';

const baseColors = {
  hackBlue: ['#B4D8FA', '#3E87CD', '#155DA1', '#0E4E8A', '#0D3F6E', '#031425'],
  indigoBlue: ['#BCD0F3', '#415F94', '#2D4979', '#253E68', '#13294E', '#061734'],
  impactGrey: ['#FAFBFC', '#EBEEF2', '#8B9199', '#666B72', '#46494F'],
};

const hackBluePalette = {
  lighter: baseColors.hackBlue[0],
  light: baseColors.hackBlue[1],
  primary: baseColors.hackBlue[2],
  dark: baseColors.hackBlue[3],
  darker: baseColors.hackBlue[4],
  text: baseColors.hackBlue[5],
};

const indigoPalette = {
  lighter: baseColors.indigoBlue[0],
  light: baseColors.indigoBlue[1],
  primary: baseColors.indigoBlue[2],
  dark: baseColors.indigoBlue[3],
  darker: baseColors.indigoBlue[4],
  text: baseColors.indigoBlue[5],
};

const impactGreyPalette = {
  lighter: baseColors.impactGrey[0],
  light: baseColors.impactGrey[1],
  midtone: baseColors.impactGrey[2],
  primary: baseColors.impactGrey[3],
  darker: baseColors.impactGrey[4],
};

const secondaryColors = {
  green: ['#BDF3C6', '#90E19E', '#4C9859', '#256830', '#154C1E', '#031D07'],
  red: ['#FFBCBF', '#D1595F', '#A62B31', '#8A1A1F', '#6B1014', '#35080A'],
  yellow: ['#FFECBD', '#FFD15F', '#FDBE21', '#DA9C03', '#A77700', '#201803'],
};

const greenPalette = {
  lighter: secondaryColors.green[0],
  light: secondaryColors.green[1],
  primary: secondaryColors.green[2],
  dark: secondaryColors.green[3],
  darker: secondaryColors.green[4],
  text: secondaryColors.green[5],
};

const redPalette = {
  lighter: secondaryColors.red[0],
  light: secondaryColors.red[1],
  primary: secondaryColors.red[2],
  dark: secondaryColors.red[3],
  darker: secondaryColors.red[4],
  text: secondaryColors.red[5],
};

const yellowPalette = {
  lighter: secondaryColors.yellow[0],
  light: secondaryColors.yellow[1],
  primary: secondaryColors.yellow[2],
  dark: secondaryColors.yellow[3],
  darker: secondaryColors.yellow[4],
  text: secondaryColors.yellow[5],
};

const typographyColors = {
  charcoal: '#272A2D',
  iron: '#4B4E52',
  ink: '#0A162A',
  slate: '#363949',
};

const colors = {
  black: '#000000',
  white: '#FFFFFF',
  ...baseColors,
  ...secondaryColors,
  ...typographyColors,
  hackBluePalette,
  indigoPalette,
  impactGreyPalette,
  greenPalette,
  redPalette,
  yellowPalette,
  text: {
    grayScale: {
      main: typographyColors.charcoal,
      support: typographyColors.iron,
    },
    coolTone: {
      main: typographyColors.ink,
      support: typographyColors.slate,
    },
  },
  // states or in other words color variants
  variants: {
    primary: hackBluePalette,
    success: greenPalette,
    error: redPalette,
    warning: yellowPalette,
    secondary: hackBluePalette,
    white: {
      primary: '#FFFFFF',
      dark: '#FFFFFF',
      lighter: indigoPalette.light,
    },
  },
};

const fontSizes = ['12px', '14px', '16px', '18px', '20px', '22px', '25px', '28px', '32px', '45px', '60px'];

const fonts = {
  chivo: fontStack(['Chivo']),
  hkGrotesk: fontStack(['HKGrotesk']),
};

const fontWeights = {
  light: 300,
  normal: 400,
  medium: 500,
  bold: 700,
};

// styling specific to the Button component
const button = {
  font: fonts.chivo,
  fontWeight: fontWeights.bold,
  fontSize: '14px',
  letterSpacing: '2px',
  // default values, based on the color palette
  default: (colorPalette) => ({
    bg: {
      default: colorPalette.primary,
      hover: colorPalette.dark,
      disabled: colorPalette.lighter,
    },
    border: {
      default: 'none',
      hover: 'none',
      disabled: 'none',
    },
    fontColor: {
      default: colors.white,
      hover: colors.white,
      disabled: colorPalette.light,
    },
    shadow: {
      default: 'none',
      hover: 'none',
      disabled: 'none',
    },
    outline: {
      bg: {
        default: colors.white,
        hover: colors.white,
        disabled: colors.white,
      },
      border: {
        default: colorPalette.primary,
        hover: colorPalette.dark,
        disabled: colorPalette.lighter,
      },
      fontColor: {
        default: colorPalette.primary,
        hover: colorPalette.dark,
        disabled: colorPalette.lighter,
      },
      shadow: {
        default: 'none',
        hover: 'none',
        disabled: 'none',
      },
    },
  }),

  // everything below overrides the default values
  primary: {
    fontColor: {
      default: colors.white,
      hover: colors.white,
      disabled: colors.white,
    },
  },
  white: {
    fontColor: {
      default: indigoPalette.primary,
      hover: indigoPalette.primary,
      disabled: indigoPalette.lighter,
    },
    shadow: {
      hover: DEFAULT_SHADOW,
    },
    outline: {
      bg: {
        default: indigoPalette.primary,
        hover: indigoPalette.primary,
        disabled: indigoPalette.primary,
      },
      border: {
        disabled: indigoPalette.light,
      },
      fontColor: {
        default: colors.white,
        hover: colors.white,
        disabled: indigoPalette.lighter,
      },
      shadow: {
        hover: DEFAULT_SHADOW,
      },
    },
  },
  secondary: {
    bg: {
      hover: hackBluePalette.darker,
      disabled: impactGreyPalette.light,
    },
    fontColor: {
      disabled: impactGreyPalette.midtone,
    },
    outline: {
      bg: {
        disabled: colors.white,
      },
      border: {
        hover: hackBluePalette.darker,
        disabled: impactGreyPalette.light,
      },
      fontColor: {
        disabled: indigoPalette.lighter,
      },
    },
  },
};


const link = {
  default: (colorPalette) => ({
    color: {
      default: colorPalette.primary,
      hover: colorPalette.dark,
    },
  }),
  secondary: {
    default: hackBluePalette.primary,
    hover: hackBluePalette.darker,
  },
};


const theme = {
  // General
  colors,
  fonts,
  fontWeights,
  fontSizes,
  button,
  link,
  DEFAULT_SHADOW,
};


export default theme;
export { colors };
