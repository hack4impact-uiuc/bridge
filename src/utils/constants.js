import * as styledSystem from 'styled-system';
import systemPropTypes from '@styled-system/prop-types';
import PropTypes from 'prop-types';


const { compose, system } = styledSystem;

// Text Decoration Prop
export const textDecoration = system({
  prop: 'textDecoration',
  cssProperty: 'text-decoration',
});
textDecoration.propTypes = PropTypes.string;


// common style props to be passed into components
export const COMMON = compose(styledSystem.space, styledSystem.layout);
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.layout,
};

// for easier access
export const TYPOGRAPHY = styledSystem.typography;
export const LAYOUT = styledSystem.layout;
export const POSITION = styledSystem.position;
export const FLEX = styledSystem.flexbox;
export const GRID = styledSystem.grid;
export const COLOR = styledSystem.color;
export const BORDER = styledSystem.border;

TYPOGRAPHY.propTypes = systemPropTypes.typography;
LAYOUT.propTypes = systemPropTypes.layout;
POSITION.propTypes = systemPropTypes.position;
FLEX.propTypes = systemPropTypes.flexbox;
GRID.propTypes = systemPropTypes.grid;
COLOR.propTypes = systemPropTypes.color;
BORDER.propTypes = systemPropTypes.border;
