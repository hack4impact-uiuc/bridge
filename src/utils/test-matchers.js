/* eslint-disable react/forbid-foreign-prop-types */
// taken from https://github.com/primer/components/blob/master/src/utils/test-matchers.js
import 'jest-styled-components';

const stringify = (d) => JSON.stringify(d, null, '  ');
/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap'];

// eslint-disable-next-line no-undef
expect.extend({
  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes));
    const expectedPropKeys = Object.keys(propNames.propTypes);
    const missing = expectedPropKeys.filter((key) => !propKeys.has(key))
      .filter((key) => !ALIAS_PROP_TYPES.includes(key));
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`,
    };
  },
});
