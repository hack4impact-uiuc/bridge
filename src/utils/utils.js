import themeGet from '@styled-system/theme-get';
import * as styledSystem from 'styled-system';
import theme from '../theme';

const { get: getKey } = styledSystem;

export const get = (key) => themeGet(key, getKey(theme, key));

export const getOrDefault = (val, def) => (val !== undefined && val !== null ? val : def);

export const lodashGet = (object, path, defaultVal) => {
  const PATH = Array.isArray(path)
    ? path
    : path.split('.').filter((i) => i.length);
  if (!PATH.length) {
    return object === undefined ? defaultVal : object;
  }
  if (object === null || object === undefined || typeof (object[PATH[0]]) === 'undefined') {
    return defaultVal;
  }
  return lodashGet(object[PATH.shift()], PATH, defaultVal);
};
