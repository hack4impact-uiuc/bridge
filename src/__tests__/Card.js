import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Button, Card } from '..';
import theme from '../theme';
import { COMMON, TYPOGRAPHY } from '../utils/constants';
import { renderJSON } from '../utils/testing';
import 'babel-polyfill'; // axe violations required babel-polyfill

const {
  Image, Content, Overline, Title, Subtitle, Body, Footer,
} = Card;

expect.extend(toHaveNoViolations);

const overline = 'Optional Overline';
const title = 'Card Title';
const subtitle = 'Optional Subtitle';
const body = 'Body';
const buttonText = 'Apply';
const exampleCard = (
  <Card>
    <Image src="https://uiuc.hack4impact.org/static/images/colored-logo.png" alt="h4i" />
    <Content>
      <Overline>{overline}</Overline>
      <Title>{title}</Title>
      <Subtitle>{subtitle}</Subtitle>
      <Body>
        {body}
      </Body>
      <Footer>
        <Button>{buttonText}</Button>
      </Footer>
    </Content>
  </Card>
);

describe('Card', () => {
  it('renders a <div>', () => {
    const json = renderJSON(<Card />);
    expect(json.type).toEqual('div');
    expect(json).toMatchSnapshot();
  });

  it('should have no axe violations', async () => {
    const { container } = render(exampleCard);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('implements system props', () => {
    expect(Card).toImplementSystemProps(COMMON);
    expect(Content).toImplementSystemProps(COMMON);
    expect(Overline).toImplementSystemProps(TYPOGRAPHY);
    expect(Overline).toImplementSystemProps(COMMON);
    expect(Title).toImplementSystemProps(TYPOGRAPHY);
    expect(Title).toImplementSystemProps(COMMON);
    expect(Subtitle).toImplementSystemProps(TYPOGRAPHY);
    expect(Subtitle).toImplementSystemProps(COMMON);
    expect(Body).toImplementSystemProps(TYPOGRAPHY);
    expect(Body).toImplementSystemProps(COMMON);
    expect(Footer).toImplementSystemProps(COMMON);
  });

  it('renders children', () => {
    render(exampleCard);
    expect(screen.getByText(overline)).toBeInTheDocument();
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(subtitle)).toBeInTheDocument();
    expect(screen.getByText(body)).toBeInTheDocument();
    expect(screen.getByText(buttonText)).toBeInTheDocument();
  });

  it('respects "width" props', () => {
    const { container } = render(<Card width={200} />);
    expect(container.firstChild).toHaveStyle('width: 200px;');
  });

  it('renders overline typography using overline styles', () => {
    const { container } = render(<Overline>content</Overline>);
    expect(container.firstChild).toHaveStyle(`font-family: ${theme.typography.overline.fontFamily}`);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.typography.overline.fontWeight}px`);
    expect(container.firstChild).toHaveStyle(`font-size: ${theme.typography.overline.fontSize}`);
    expect(container.firstChild).toHaveStyle(`letter-spacing: ${theme.typography.overline.letterSpacing}`);
    expect(container.firstChild).toHaveStyle(`line-height: ${theme.typography.overline.lineHeight}`);
  });

  it('renders title typography using h3 styles', () => {
    const { container } = render(<Title>content</Title>);
    expect(container.firstChild).toHaveStyle(`font-family: ${theme.typography.h3.fontFamily}`);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.typography.h3.fontWeight}px`);
    expect(container.firstChild).toHaveStyle(`font-size: ${theme.typography.h3.fontSize}`);
    expect(container.firstChild).toHaveStyle(`letter-spacing: ${theme.typography.h3.letterSpacing}`);
    expect(container.firstChild).toHaveStyle(`line-height: ${theme.typography.h3.lineHeight}`);
  });

  it('renders subtitle typography using subtitle styles', () => {
    const { container } = render(<Subtitle>content</Subtitle>);
    expect(container.firstChild).toHaveStyle(`font-family: ${theme.typography.subtitle.fontFamily}`);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.typography.subtitle.fontWeight}px`);
    expect(container.firstChild).toHaveStyle(`font-size: ${theme.typography.subtitle.fontSize}`);
    expect(container.firstChild).toHaveStyle(`letter-spacing: ${theme.typography.subtitle.letterSpacing}`);
    expect(container.firstChild).toHaveStyle(`line-height: ${theme.typography.subtitle.lineHeight}`);
  });

  it('renders body typography using body styles', () => {
    const { container } = render(<Body>content</Body>);
    expect(container.firstChild).toHaveStyle(`font-family: ${theme.typography.body.fontFamily}`);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.typography.body.fontWeight}px`);
    expect(container.firstChild).toHaveStyle(`font-size: ${theme.typography.body.fontSize}`);
    expect(container.firstChild).toHaveStyle(`letter-spacing: ${theme.typography.body.letterSpacing}`);
    expect(container.firstChild).toHaveStyle(`line-height: ${theme.typography.body.lineHeight}`);
  });

  it('renders entire card properly', () => {
    expect(renderJSON(exampleCard)).toMatchSnapshot();
  });
});
