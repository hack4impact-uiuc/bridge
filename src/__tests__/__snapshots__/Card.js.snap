// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders a <div> 1`] = `
.c0 {
  display: inline-block;
  vertical-align: top;
  width: 471px;
  background-color: #FFFFFF;
  border-radius: 8px;
  box-shadow: 0 2px 5px 0 rgba(0,0,0,0.24);
}

.c0:hover {
  box-shadow: 0 2px 16px 0 rgba(0,0,0,0.24);
}

<div
  className="c0"
/>
`;

exports[`Card renders entire card properly 1`] = `
.c8 {
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow: hidden;
  border-radius: 13px;
  font-family: Chivo;
  font-weight: 700px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  font-size: 14px;
  line-height: 20px;
  padding: 12px 16px;
  display: inline-block;
  text-align: center;
  vertical-align: middle;
  background: #155DA1;
  border: none;
  color: #FFFFFF;
}

.c8:hover:not([disabled]) {
  background: #0E4E8A;
  border: none;
  box-shadow: none;
  color: #FFFFFF;
  cursor: pointer;
}

.c8:disabled {
  background: #B4D8FA;
  border: none;
  box-shadow: none;
  color: #FFFFFF;
}

.c1 {
  width: 100%;
}

.c2 {
  padding: 28px;
}

.c3 {
  font-family: HKGrotesk;
  font-weight: 700px;
  font-size: 12px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  line-height: 16px;
  text-transform: uppercase;
  color: #0A162A;
  margin-bottom: 8px;
}

.c4 {
  font-family: Chivo;
  font-weight: 400px;
  font-size: 32px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 48px;
  color: #0A162A;
}

.c5 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 20px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 28px;
  color: #0A162A;
  margin-top: 8px;
}

.c6 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 16px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 24px;
  color: #0A162A;
  margin-top: 16px;
}

.c7 {
  margin-top: 24px;
}

.c0 {
  display: inline-block;
  vertical-align: top;
  width: 471px;
  background-color: #FFFFFF;
  border-radius: 8px;
  box-shadow: 0 2px 5px 0 rgba(0,0,0,0.24);
}

.c0:hover {
  box-shadow: 0 2px 16px 0 rgba(0,0,0,0.24);
}

<div
  className="c0"
>
  <img
    alt="h4i"
    className="c1"
    src="https://uiuc.hack4impact.org/static/images/colored-logo.png"
  />
  <div
    className="c2"
  >
    <div
      className="c3"
    >
      Optional Overline
    </div>
    <div
      className="c4"
    >
      Card Title
    </div>
    <div
      className="c5"
    >
      Optional Subtitle
    </div>
    <div
      className="c6"
    >
      Body
    </div>
    <div
      className="c7"
    >
      <button
        className="c8"
        type="medium"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
